packages/daemon/src/app/controller/plugin.ts

```diff
@@ -1,68 +1,117 @@
+import { BaseController } from './base';
 import { PluginManager } from '../../service/plugin';
 import ServerSentEmitter from '../../utils/emitter';
+import { PluginInstallingResp } from '../../interface';
 import Debug from 'debug';
 const debug = Debug('daemon.app.plugin');
 
 @provide()
 @controller('/plugin')
-export class PluginController {
-  @inject()
-  ctx: Context;
+export class PluginController extends BaseController {
 
   @inject('pluginManager')
   pluginManager: PluginManager;
 
-  @get('/install')
+  // TODO(feely): check if the plugin has been installed
+  /**
+   * create plugin and install it
+   */
+  @post()
   public async install() {
+    const { name, pyIndex } = this.ctx.request.body;
+    let response: PluginInstallingResp;
+    debug(`checking info: ${name}.`);
+    try {
+      response = await this.pluginManager.installByName(name, pyIndex, false);
+      this.success(response);
+    } catch (err) {
+      this.fail(`plugin installation failed with error: ${err.message}`);
+    }
+  }
+
+  /**
+   * reinstall plugin
+   */
+  @put()
+  public async reinstall() {
     const { name, pyIndex } = this.ctx.query;
-    const sse = new ServerSentEmitter(this.ctx);
+    let response: PluginInstallingResp;
+    debug(`checking info: ${name}.`);
     try {
-      debug(`checking info: ${name}.`);
-      const pkg = await this.pluginManager.fetch(name);
-      sse.emit('info', pkg);
+      response = await this.pluginManager.installByName(name, pyIndex, true);
+      this.success(response);
+    } catch (err) {
+      this.fail(`plugin installation failed with error: ${err.message}`);
+    }
+  }
 
-      debug(`installing ${name}.`);
-      const plugin = await this.pluginManager.findOrCreateByPkg(pkg);
-      try {
-        await this.pluginManager.install(pkg, pyIndex);
-        sse.emit('installed', pkg);
-      } catch (err) {
-        await this.pluginManager.removeById(plugin.id);
-        throw err;
+  /**
+   * delete plugin by name
+   */
+  @del('/:id')
+  public async remove() {
+    try {
+      if (typeof this.ctx.params.id === 'string' && this.ctx.params.id) {
+        const plugin = await this.pluginManager.findById(this.ctx.params.id);
+        await this.pluginManager.uninstall(plugin.name);
+        this.success(undefined, 204);
+      } else {
+        // not implemented
+        this.fail('no name value found', 400);
       }
     } catch (err) {
-      sse.emit('error', err?.message);
-    } finally {
-      sse.finish();
+      this.fail(err.message, 404);
     }
   }
-
-  @get('/uninstall')
-  public async uninstall() {
-    await this.pluginManager.uninstall(this.ctx.query.name);
-    successRes(this.ctx, {});
+  /**
+   * delete all plugins
+   */
+  @del()
+  public async removeAll() {
+    try {
+      const plugins = await this.pluginManager.list();
+      for (const plugin of plugins) {
+        await this.pluginManager.uninstall(plugin.name);
```

Comment:
May I uninstall them in parallel?

Reply:
If we do uninstall in parallel, we should return code 202, and add a new status `removing` for plugin.